pip/qsharp/_qsharp.py

```diff
@@ -326,9 +363,68 @@ def estimate(
             params = [params.as_dict()]
     elif isinstance(params, dict):
         params = [params]
-    return EstimatorResult(
-        json.loads(get_interpreter().estimate(entry_expr, json.dumps(params)))
-    )
+    return params
+
+
+def estimate(
+    estimate_input: Union[str, "QuantumCircuit"],
+    params: Optional[
+        Union[Dict[str, Any], List[Dict[str, Any]], EstimatorParams]
+    ] = None,
+    **options,
+) -> EstimatorResult:
+    """
+    Estimates resources for the quantum algorithm.
+
+    :param estimate_input: The Q# entry expression or `QuantumCircuit`.
+    :param params: The parameters to configure physical estimation.
+    :**options: Additional options for the `QiskitCircuit` input.
+            - Any options for the transpiler, exporter, or Qiskit passes
+                configuration. Defaults to backend config values. Common
+                values include: 'optimization_level', 'basis_gates',
+                'includes', 'search_path'.
+    :raises `QasmError`: If there is an error generating or parsing QASM.
+    :returns `EstimatorResult`: The estimated resources.
+    """
+
+    ipython_helper()
+    params = _coerce_estimator_params(params)
+    param_str = json.dumps(params)
+    if isinstance(estimate_input, str):
+        res_str = get_interpreter().estimate(estimate_input, param_str)
+        res = json.loads(res_str)
+        return EstimatorResult(res)
+    else:
+
+        def issubclassofbyname(ty: type, ty_name: str):
+            if ty.__name__ == ty_name:
+                return True
+            for base_ty in ty.__bases__:
+                if issubclassofbyname(base_ty, ty_name):
+                    return True
+            return False
+
+        if issubclassofbyname(type(estimate_input), "QuantumCircuit"):
+            # we know the ty name is QuantumCircuit, but if we can't import Qiskit
+            # then we know it is a name collision and invalid input type
+            try:
+                # We delay import Qisit to avoid the user needing to install it
+                # if they are not using the `estimate` function with a `QuantumCircuit`.
+
+                # if Qiskit isn't installed, this import will fail
+                from qsharp.interop.qiskit.backends import ResourceEstimatorBackend
+
+                backend = ResourceEstimatorBackend(
+                    target_profile=TargetProfile.Unrestricted
+                )
+                job = backend.run(estimate_input, params=params, **options)
+                return job.result()
+            except ImportError as ex:
+                message = "Unsupported input type."
+                raise ValueError(message) from ex
```

Comment:
would it make sense to include something about failing to import qiskit here, so users know they've hit a recognized path that we can't support? It seems a little odd that the general error below includes more information than this one. It's possible the second line here that has `from ex` includes the detailed import error, in which case sorry for the noise!